[C++] default(T) initializer list may lead to ambiguous assignment and prevent compilation

`default(T)` does not use fully qualified C++ constructors as described in https://nim-lang.org/docs/manual_experimental.html#constructor-pragma. It uses initializer lists.

This is particularly problematic as it is used in setLen for sequences see line 268
https://github.com/nim-lang/Nim/blob/4c8052a45bc12b0dc1114ca606d142d33495d8f2/lib/system/seqs_v2.nim#L253-L269


Repro

```C++
// ambiguous_assign.h
#ifndef AMBIGUOUS_ASSIGN_H
#define AMBIGUOUS_ASSIGN_H

// Minimal struct with multiple assignment operators that make {} ambiguous
struct AmbiguousAssign {
  int x;
  const char* y;
  
  AmbiguousAssign(): x(0), y(nullptr) {}
  AmbiguousAssign(int x, const char* y): x(x), y(y) {}
  
  // Assignment from int
  AmbiguousAssign& operator=(int v) {
    x = v;
    y = nullptr;
    return *this;
  }
  
  // Assignment from const char*
  AmbiguousAssign& operator=(const char* s) {
    x = 0;
    y = s;
    return *this;
  }
  
  // Copy assignment
  AmbiguousAssign& operator=(const AmbiguousAssign& other) {
    x = other.x;
    y = other.y;
    return *this;
  }
};

#endif
```
```nim
# ambiguous_assign.nim
import std/os

const HeaderPath* = currentSourcePath.parentDir()

{.passC: "-I" & HeaderPath.}

type
  AmbiguousAssign* {.importcpp: "AmbiguousAssign", header: "ambiguous_assign.h".} = object
    x*: cint
    y*: cstring

proc main =
  # This triggers setLen which uses default(T) -> {} in C++
  # The {} is ambiguous between the multiple assignment operators
  var s = newSeq[AmbiguousAssign](3)
  echo "Created sequence"

main()
```

Error
```
[...]/nim_bugs/ambiguous_assign.nim(15, 7) Hint: 's' is declared but not used [XDeclaredButNotUsed]
CC: system/exceptions.nim
CC: std/private/digitsutils.nim
CC: system.nim
CC: times.nim
CC: std/envvars.nim
CC: std/cmdline.nim
CC: ambiguous_assign.nim
[...]/nim_bugs/nimcache/ambiguous_assign/@psystem.nim.cpp: In function ‘void setLen__ambiguous95assign_u53(tySequence__K7NZ8Ti2277HeyrLi50SRg&, NI)’:
[...]/nim_bugs/nimcache/ambiguous_assign/@psystem.nim.cpp:3451:67: error: ambiguous overload for ‘operator=’ (operand types are ‘AmbiguousAssign’ and ‘<brace-enclosed initializer list>’)
 3451 |                                         (*(*xu_1).p).data[i_1] = {};
      |                                                                   ^
[...]/nim_bugs/nimcache/ambiguous_assign/@psystem.nim.cpp:3451:67: note: there are 3 candidates
In file included from [...]/nim_bugs/nimcache/ambiguous_assign/@psystem.nim.cpp:14:
[...]/nim_bugs/ambiguous_assign.h:13:20: note: candidate 1: ‘AmbiguousAssign& AmbiguousAssign::operator=(int)’
   13 |   AmbiguousAssign& operator=(int v) {
      |                    ^~~~~~~~
[...]/nim_bugs/ambiguous_assign.h:20:20: note: candidate 2: ‘AmbiguousAssign& AmbiguousAssign::operator=(const char*)’
   20 |   AmbiguousAssign& operator=(const char* s) {
      |                    ^~~~~~~~
[...]/nim_bugs/ambiguous_assign.h:27:20: note: candidate 3: ‘AmbiguousAssign& AmbiguousAssign::operator=(const AmbiguousAssign&)’
   27 |   AmbiguousAssign& operator=(const AmbiguousAssign& other) {
      |                    ^~~~~~~~
Error: execution of an external compiler program 'g++ -c -std=gnu++17 -funsigned-char  -w -fmax-errors=3 -fpermissive -pthread -I[...]/nim_bugs/nimcache/ambiguous_assign/@psystem.nim.cpp' failed with exit code: 1
```

--- a/compiler/ccgtypes.nim
+++ b/compiler/ccgtypes.nim
@@ -696,6 +696,24 @@ proc genCppInitializer(m: BModule, prc: BProc; typ: PType; didGenTemp: var bool)
       if prc == nil:
         assert p.blocks.len == 0, "BProc belongs to a struct doesnt have blocks"
 
+# Unlike genCppInitializer which returns just the braced value list (e.g. "{a, b}"),
+# genCppConstructorExpr returns a full type-prefixed expression (e.g. "Foo(a, b)").
+# This is used when a standalone construction expression is needed — e.g. on the
+# right-hand side of an assignment — whereas genCppInitializer is used in variable
+# declarations where the type is already written separately before the initializer.
+proc genCppConstructorExpr(m: BModule, prc: BProc; typ: PType; didGenTemp: var bool): Snippet =
+  var params = ""
+  if typ.itemId in m.g.graph.initializersPerType:
+    let call = m.g.graph.initializersPerType[typ.itemId]
+    if call != nil:
+      var p = prc
+      if p == nil:
+        p = BProc(module: m)
+      params = genCppParamsForCtor(p, call, didGenTemp)
+      if prc == nil:
+        assert p.blocks.len == 0, "BProc belongs to a struct doesnt have blocks"
+  result = getTypeDesc(m, typ, dkVar) & "(" & params & ")"
+
 proc genRecordFieldsAux(m: BModule; n: PNode,
                         rectype: PType,
                         check: var IntSet; result: var Builder; unionPrefix = "") =

--- a/compiler/ccgtypes.nim
+++ b/compiler/ccgtypes.nim
@@ -696,24 +696,6 @@ proc genCppInitializer(m: BModule, prc: BProc; typ: PType; didGenTemp: var bool)
       if prc == nil:
         assert p.blocks.len == 0, "BProc belongs to a struct doesnt have blocks"
 
-# Unlike genCppInitializer which returns just the braced value list (e.g. "{a, b}"),
-# genCppConstructorExpr returns a full type-prefixed expression (e.g. "Foo(a, b)").
-# This is used when a standalone construction expression is needed — e.g. on the
-# right-hand side of an assignment — whereas genCppInitializer is used in variable
-# declarations where the type is already written separately before the initializer.
-proc genCppConstructorExpr(m: BModule, prc: BProc; typ: PType; didGenTemp: var bool): Snippet =
-  var params = ""
-  if typ.itemId in m.g.graph.initializersPerType:
-    let call = m.g.graph.initializersPerType[typ.itemId]
-    if call != nil:
-      var p = prc
-      if p == nil:
-        p = BProc(module: m)
-      params = genCppParamsForCtor(p, call, didGenTemp)
-      if prc == nil:
-        assert p.blocks.len == 0, "BProc belongs to a struct doesnt have blocks"
-  result = getTypeDesc(m, typ, dkVar) & "(" & params & ")"
-
 proc genRecordFieldsAux(m: BModule; n: PNode,
                         rectype: PType,
                         check: var IntSet; result: var Builder; unionPrefix = "") =
@@ -2130,3 +2112,21 @@ proc genTypeSection(m: BModule, n: PNode) =
       discard getTypeDescAux(m, s.typ, intSet, descKindFromSymKind(s.kind))
       if m.g.generatedHeader != nil:
         discard getTypeDescAux(m.g.generatedHeader, s.typ, intSet, descKindFromSymKind(s.kind))
+
+# Unlike genCppInitializer which returns just the braced value list (e.g. "{a, b}"),
+# genCppConstructorExpr returns a full type-prefixed expression (e.g. "Foo(a, b)").
+# This is used when a standalone construction expression is needed — e.g. on the
+# right-hand side of an assignment — whereas genCppInitializer is used in variable
+# declarations where the type is already written separately before the initializer.
+proc genCppConstructorExpr(m: BModule, prc: BProc; typ: PType; didGenTemp: var bool): Snippet =
+  var params = ""
+  if typ.itemId in m.g.graph.initializersPerType:
+    let call = m.g.graph.initializersPerType[typ.itemId]
+    if call != nil:
+      var p = prc
+      if p == nil:
+        p = BProc(module: m)
+      params = genCppParamsForCtor(p, call, didGenTemp)
+      if prc == nil:
+        assert p.blocks.len == 0, "BProc belongs to a struct doesnt have blocks"
+  result = getTypeDesc(m, typ, dkVar) & "(" & params & ")"

--- a/compiler/cgen.nim
+++ b/compiler/cgen.nim
@@ -526,7 +526,7 @@ proc resetLoc(p: BProc, loc: var TLoc) =
   if isImportedCppType(typ):
     var didGenTemp = false
     let rl = rdLoc(loc)
-    let init = genCppInitializer(p.module, p, typ, didGenTemp)
+    let init = genCppConstructorExpr(p.module, p, typ, didGenTemp)
     p.s(cpsStmts).addAssignment(rl, init)
     return
   if optSeqDestructors in p.config.globalOptions and typ.kind in {tyString, tySequence}:

--- a/tests/cpp/tcpp_default_ctor_assignment.h
+++ b/tests/cpp/tcpp_default_ctor_assignment.h
@@ -1,0 +1,30 @@
+#ifndef TCPP_DEFAULT_CTOR_ASSIGNMENT_H
+#define TCPP_DEFAULT_CTOR_ASSIGNMENT_H
+
+struct AmbiguousAssign {
+  int x;
+  const char* y;
+
+  AmbiguousAssign(): x(0), y(nullptr) {}
+  AmbiguousAssign(int x, const char* y): x(x), y(y) {}
+
+  AmbiguousAssign& operator=(int v) {
+    x = v;
+    y = nullptr;
+    return *this;
+  }
+
+  AmbiguousAssign& operator=(const char* s) {
+    x = 0;
+    y = s;
+    return *this;
+  }
+
+  AmbiguousAssign& operator=(const AmbiguousAssign& other) {
+    x = other.x;
+    y = other.y;
+    return *this;
+  }
+};
+
+#endif

--- a/tests/cpp/tcpp_default_ctor_assignment.nim
+++ b/tests/cpp/tcpp_default_ctor_assignment.nim
@@ -1,0 +1,14 @@
+discard """
+  cmd: "nim cpp $file"
+"""
+
+type
+  AmbiguousAssign {.importcpp, header: "tcpp_default_ctor_assignment.h".} = object
+    x: cint
+    y: cstring
+
+proc main =
+  var xs = newSeq[AmbiguousAssign](3)
+  doAssert xs.len == 3
+
+main()

